### PR TITLE
snax: Add wrapper and template generation

### PR DESCRIPTION
**Pull request by rgantonio**
This PR imports the wrappers made from the [snax-dev ](https://github.com/KULeuven-MICAS/snax-dev) repo.

To make this PR understandable to all, this PR makes wrappers to accommodate the generated files in chisel and combining with SystemVerilog files. The wrapper is necessary because Chisel files generate unpacked signals one by one making it hard to manually hook up signals. So this PR makes those wrappers that automatically hook up signals making it easier for people to use. To understand better check the figure below:

image

Let's go through this one by one:
1. The red boxes pertain to the chisel-generated modules. In this PR we have the following:
- We first need to create parameters that chisel files need for the streamer and CSR manager. This includes the streamer characteristics like how many FIFOs? What is the depth of each FIFO? What is the data width? How many readers and writers? And so on. So the `wrappergen.py` generates these parameters. The parameters are also using templates from the `hw/templates` directory.
- Next, after creating the parameters, we invoke Chisel to use those parameters and generate the streamers and CSR manager. The `chiselgen.py` invokes the Chisel to generate these.

2. The blue boxes pertain to the wrappers because we need to combine signals that are easier for people to use. Hence from the unpacked signals of the chisel-generated files, we combine these into packed signals. In this PR we have:

- Several templates inside the `hw/template` directory for these things. When you scan them, they're literally just organizing signals that can vary depending on the configuration set inside the `target/snitch_cluster/cfg` directory.
- So basically, the `wrappergen` grabs the config and template files to create the wrappers.
- There are three major wrappers: `streamer_wrapper` for the chisel-generated streamer, `csr_manager_wrapper` for the chisel-generated CSR manager, and finally the `top_level_wrapper` that organizes and encapsulates all the modules.

The white boxes are external signals. Take note that the accelerator is assumed to be developed in System Verilog.

There's a CI added to do sanity-checking on these wrappers.

TODO left for this PR:

- [x] Streamer wrapper
- [x] Csr wrapper
- [x] Top level wrapper
- [x] Generator script
- [x] Test to check if generation works
- [x] Add to CI to check if generation works
- [x] Test to check if chisel params work
- [x] Add to CI to check if chisel params work

### CODE REVIEW

File: hw/snax_wrappers/csrman_param_gen.scala.tpl [hunk 1]
```diff
@@ -0,0 +1,33 @@
+<%
+  import math
+
+  tcdm_data_width = cfg["tcdmDataWidth"]
```
Comment (xiaoling-yi) [inline hw/snax_wrappers/csrman_param_gen.scala.tpl:4]: We don't need these parameter here?

Comment (rgantonio) [reply hw/snax_wrappers/csrman_param_gen.scala.tpl]: Done!

File: hw/templates/snax_accelerator_top_wrapper.sv.tpl [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+  //-----------------------------
+  // Streamer Wrapper
+  //-----------------------------
+  ${cfg["tag_name"]}_streamer_wrapper #(
+    .NarrowDataWidth          ( NarrowDataWidth         ),
+    .TCDMDepth                ( TCDMDepth               ),
+    .TCDMReqPorts             ( TCDMReqPorts            ),
+    .TCDMSize                 ( TCDMSize                ),
+    .TCDMAddrWidth            ( TCDMAddrWidth           )
+  ) i_streamer_wrapper (
```
Comment (xiaoling-yi) [inline hw/templates/snax_accelerator_top_wrapper.sv.tpl:223]: Maybe changing `i_streamer_wrapper` to `i_${cfg["tag_name"]}_streamer_wrapper` is better? Otherwise if there are multiple accelerators, this `i_streamer_wrapper` module will cause multiple definition error

File: hw/templates/snax_csrman_wrapper.sv.tpl [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    .io_csr_config_in_rsp_bits_data   ( csr_rsp_data_o      ),
+    .io_csr_config_in_rsp_valid       ( csr_rsp_valid_o     ),
+    .io_csr_config_in_rsp_ready       ( csr_rsp_ready_i     ),	
+
+    //-----------------------------
+    // CSR configurations
+    //-----------------------------
+% for i in range(cfg["snax_acc_num_csr"]):
+    .io_csr_config_out_bits_${i}      ( csr_reg_set_o[${i}] ),
+% endfor
```
Comment (xiaoling-yi) [inline hw/templates/snax_csrman_wrapper.sv.tpl:68]: The actual output ports here is `cfg["snax_acc_num_csr"] - 1`. The last csr output is omitted in the generated system verilog file because the last is status csr and no need to output to the accelerator.

File: hw/templates/snax_streamer_wrapper.sv.tpl [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+  input  logic clk_i,
+  input  logic rst_ni,
+
+  //-----------------------------
+  // Accelerator ports
+  //-----------------------------
+  // Input ports from accelerator to streamer
+  input  logic [ NumInputPorts-1:0][StreamerDataWidth-1:0] acc2stream_data_i,
+  input  logic [ NumInputPorts-1:0]                        acc2stream_valid_i,
+  output logic [ NumInputPorts-1:0]                        acc2stream_ready_o,
```
Comment (xiaoling-yi) [inline hw/templates/snax_streamer_wrapper.sv.tpl:43]: Bug!! Same issue. What is the exact meaning of `NumInputPorts` and `NumOutputPorts`. They have double meaning now and will cause functional errors!

File: hw/templates/snax_accelerator_top_wrapper.sv.tpl [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+// SPDX-License-Identifier: SHL-0.51
+
+<%
+  num_input_ports = len(cfg["snax_streamer_cfg"]["data_reader_params"]["tcdm_ports_num"])
+  num_output_ports = len(cfg["snax_streamer_cfg"]["data_writer_params"]["tcdm_ports_num"])
+  num_tcdm_ports = num_input_ports + num_output_ports
+
+  # We make the assumption that all reader and writers
+  # Have the same data widths
+  stream_data_width = cfg["snax_streamer_cfg"]["data_reader_params"]["element_width"][0]
```
Comment (xiaoling-yi) [inline hw/templates/snax_accelerator_top_wrapper.sv.tpl:12]: But the output port of the accelerator is larger than the input port for normal accelerators? Is that the cause of the ALU accelerator?

File: hw/templates/snax_accelerator_top_wrapper.sv.tpl [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+  logic                    acc_csr_reg_set_valid;
+  logic                    acc_csr_reg_set_ready;
+
+  //-------------------------------
+  // MUX and DEMUX for control signals
+  // That separate between streamer CSR
+  // and accelerator CRS
+  //-------------------------------
+  csr_mux_demux #(
+    .AddrSelOffSet        ( 8                     ),
```
Comment (xiaoling-yi) [inline hw/templates/snax_accelerator_top_wrapper.sv.tpl:113]: Why the CSR address offset is always 8? What if the accelerator has more than 8 CSRs?

File: hw/templates/snax_streamer_wrapper.sv.tpl [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    //-----------------------------
+    // Clocks and reset
+    //-----------------------------
+    .clock ( clk_i   ),
+    .reset ( ~rst_ni ),
+
+    //-----------------------------
+    // Accelerator ports
+    //-----------------------------
+% for idx, dw in enumerate(cfg["snax_streamer_cfg"]["fifo_writer_params"]["fifo_width"]):
```
Comment (xiaoling-yi) [inline hw/templates/snax_streamer_wrapper.sv.tpl:113]: Bug!! Maybe you mean something like `["fifo_num"]`, which equals to `len(cfg["snax_streamer_cfg"]["fifo_writer_params"]["fifo_width"])`?

File: hw/templates/snax_streamer_wrapper.sv.tpl [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+    .io_data_accelerator2streamer_data_${idx}_bits  ( acc2stream_data_i  [${idx}] ),
+    .io_data_accelerator2streamer_data_${idx}_valid ( acc2stream_valid_i [${idx}] ),
+    .io_data_accelerator2streamer_data_${idx}_ready ( acc2stream_ready_o [${idx}] ),
+
+% endfor
+% for idx, dw in enumerate(cfg["snax_streamer_cfg"]["fifo_reader_params"]["fifo_width"]):
```
Comment (xiaoling-yi) [inline hw/templates/snax_streamer_wrapper.sv.tpl:119]: Bug too!!

File: hw/templates/stream_param_gen.scala.tpl [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+  * loop bound addrWidth - the address width stationarity - accelerator
+  * stationarity feature for each data mover (data reader and data writer)
+  */
+
+// Streamer parameters
+object StreamerParametersGen extends CommonParams {
+  def temporalAddrGenUnitParams: TemporalAddrGenUnitParams =
+    TemporalAddrGenUnitParams(
+      loopDim = ${cfg["snax_streamer_cfg"]["temporal_addrgen_unit_params"]["loop_dim"]},
+      loopBoundWidth = ${cfg["snax_streamer_cfg"]["temporal_addrgen_unit_params"]["loop_bound_width"]},
```
Comment (xiaoling-yi) [inline hw/templates/stream_param_gen.scala.tpl:51]: What about we fix this `loopBoundWidth` to 32? As each `loopBoundWidth` is set by on CSR and one CSR is 32 bits.

File: hw/templates/stream_param_gen.scala.tpl [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+      fifoWidth = fifoWriterParams(${idx}).width
+    )${', ' if not loop.last else ''}
+% endfor
+  )
+  def stationarity = Seq(${list_elem('stationarity')})
+}
+
+object StreamerTopGen {
+  def main(args: Array[String]) : Unit = {
+    val outPath = args.headOption.getOrElse("../../../../rtl/.")
```
Comment (xiaoling-yi) [inline hw/templates/stream_param_gen.scala.tpl:105]: Same as mentioned before!

File: target/snitch_cluster/cfg/snax-alu.hjson [hunk 11]
```diff
@@ -0,0 +1,190 @@
+// Copyright 2023 ETH Zurich and University of Bologna.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
```
Comment (xiaoling-yi) [inline target/snitch_cluster/cfg/snax-alu.hjson:3]: Maybe KU Leuven lic?

File: target/snitch_cluster/cfg/snax-alu.hjson [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+        xf16: true,
+        xf16alt: true,
+        xf8: true,
+        xf8alt: true,
+        xfdotp: true,
+        xfvec: true,
+        snax_acc_cfg: {
+            snax_acc_name: "snax_alu_core1"
+            snax_module: "snax_alu_wrapper",
+            snax_tcdm_ports: 4,
```
Comment (xiaoling-yi) [inline target/snitch_cluster/cfg/snax-alu.hjson:93]: Is the tcdm port number really 4? It is 12 if I infer from the streamer cfg below.

File: target/snitch_cluster/cfg/snax-alu.hjson [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+        xf16: true,
+        xf16alt: true,
+        xf8: true,
+        xf8alt: true,
+        xfdotp: true,
+        xfvec: true,
+        snax_acc_cfg: {
+            snax_acc_name: "snax_alu_core2"
+            snax_module: "snax_alu_wrapper",
+            snax_tcdm_ports: 4,
```
Comment (xiaoling-yi) [inline target/snitch_cluster/cfg/snax-alu.hjson:121]: Same concern as above.

File: util/chiselgen/chiselgen.py [hunk 14]
```diff
@@ -17,28 +17,32 @@ def main():
...
         required=True,
-        help="The specific parameter class for the streamer \
-        to generate the system verilog file",
+        help="Parameter class name",
     )
     parser.add_argument(
-        "--outdir",
-        "-o",
+        "--chisel_path",
+        type=str,
```
Comment (xiaoling-yi) [inline util/chiselgen/chiselgen.py:27]: If it is a path, then the type should be `pathlib.Path`?

**Review by jorendumoulin — COMMENTED** [review]
Overall this looks good!
I'll have a look if i can resolve the Chisel build errors

File: hw/templates/snax_streamer_wrapper.sv.tpl [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+module ${cfg["tag_name"]}_streamer_wrapper #(
+  // Parameters related to TCDM
+  parameter int unsigned NarrowDataWidth   = ${cfg["tcdm_data_width"]},
+  parameter int unsigned TCDMDepth         = ${cfg["tcdm_depth"]},
+  parameter int unsigned TCDMReqPorts      = ${num_tcdm_ports},
+  parameter int unsigned NrBanks           = ${cfg["num_banks"]},
+  parameter int unsigned TCDMSize          = NrBanks * TCDMDepth * (NarrowDataWidth/8),
+  parameter int unsigned TCDMAddrWidth     = $clog2(TCDMSize),
+  // Parameters related to streamers
+  // Touch at your own risk!
```
Comment (JosseVanDelm) [inline hw/templates/snax_streamer_wrapper.sv.tpl:26]: nit: I don't understand this warning very well. Maybe we should put a warning on top of all template files that they are generated by a template instead?

**Review by JosseVanDelm — COMMENTED** [review]
Overall this PR looks good to me, I just have some small comments.
Also thanks for the thorough reviews @xiaoling-yi  and @jorendumoulin !

**Review by JosseVanDelm — COMMENTED** [review]
Thanks for addressing my comments @rgantonio !

File: hw/templates/snax_accelerator_top_wrapper.sv.tpl (hunk 5 above)
Comment (rgantonio) [reply hw/templates/snax_accelerator_top_wrapper.sv.tpl:13]: After discussing with @xiaoling-yi  we use this:

`stream_data_width = cfg["snax_streamer_cfg"]["fifo_reader_params"]["fifo_width"][0]`

File: hw/templates/snax_streamer_wrapper.sv.tpl (hunk 7 above)
Comment (rgantonio) [reply hw/templates/snax_streamer_wrapper.sv.tpl:117]: Fixed!

**Review by xiaoling-yi — APPROVED** [review]
